pkg/remote/crd_upgrade.go

```diff
@@ -21,6 +25,136 @@ func PatchCRD(ctx context.Context, clnt client.Client, crd *v1extensions.CustomR
+	if kyma.Annotations == nil {
+		kyma.Annotations = make(map[string]string)
+	}
+	annotation := getAnnotation(crd, crdType)
+	kyma.Annotations[annotation] = strconv.FormatInt(crd.Generation, 10)
+}
+
+func getAnnotation(crd *v1extensions.CustomResourceDefinition, crdType CrdType) string {
+	return fmt.Sprintf("%s-%s-crd-generation", strings.ToLower(crd.Spec.Names.Kind), strings.ToLower(string(crdType)))
+}
+
+func SyncCrdsAndUpdateKymaAnnotations(ctx context.Context, kyma *v1beta2.Kyma,
+	runtimeClient Client, controlPlaneClient Client) (bool, error) {
+	kcpKymaCrd, skrKymaCrd, err := fetchCrds(ctx, controlPlaneClient, runtimeClient, v1beta2.KymaKind.Plural())
+	if err != nil {
+		return false, err
+	}
+	kymaCrdUpdated, err := updateRemoteCRD(ctx, kyma, runtimeClient, skrKymaCrd, kcpKymaCrd)
+	if err != nil {
+		return false, err
+	}
+	if kymaCrdUpdated {
+		err = runtimeClient.Get(
+			ctx, client.ObjectKey{
+				Name: fmt.Sprintf("%s.%s", v1beta2.KymaKind.Plural(), v1beta2.GroupVersion.Group),
+			}, skrKymaCrd,
+		)
+		if err != nil {
+			return false, err
+		}
+		updateKymaAnnotations(kyma, kcpKymaCrd, KCP)
+		updateKymaAnnotations(kyma, skrKymaCrd, SKR)
+	}
+
+	kcpModuleTemplateCrd, skrModuleTemplateCrd, err := fetchCrds(ctx, controlPlaneClient, runtimeClient,
+		v1beta2.ModuleTemplateKind.Plural())
+	if err != nil {
+		return false, err
+	}
+	moduleTemplateCrdUpdated, err := updateRemoteCRD(ctx, kyma, runtimeClient, skrModuleTemplateCrd, kcpModuleTemplateCrd)
+	if err != nil {
+		return false, err
+	}
+	if moduleTemplateCrdUpdated {
+		err = runtimeClient.Get(
+			ctx, client.ObjectKey{
+				Name: fmt.Sprintf("%s.%s", v1beta2.ModuleTemplateKind.Plural(), v1beta2.GroupVersion.Group),
+			}, skrModuleTemplateCrd,
+		)
+		if err != nil {
+			return false, err
+		}
+		updateKymaAnnotations(kyma, kcpModuleTemplateCrd, KCP)
+		updateKymaAnnotations(kyma, skrModuleTemplateCrd, SKR)
+	}
+
+	return kymaCrdUpdated || moduleTemplateCrdUpdated, nil
+}
+
+func fetchCrds(ctx context.Context, controlPlaneClient Client, runtimeClient Client, plural string) (
+	*v1extensions.CustomResourceDefinition, *v1extensions.CustomResourceDefinition, error) {
+	crd := &v1extensions.CustomResourceDefinition{}
+	crdFromRuntime := &v1extensions.CustomResourceDefinition{}
+	var err error
+	err = controlPlaneClient.Get(
+		ctx, client.ObjectKey{
+			// this object name is derived from the plural and is the default kustomize value for crd namings, if the CRD
+			// name changes, this also has to be adjusted here. We can think of making this configurable later
+			Name: fmt.Sprintf("%s.%s", plural, v1beta2.GroupVersion.Group),
+		}, crd,
+	)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = runtimeClient.Get(
+		ctx, client.ObjectKey{
+			Name: fmt.Sprintf("%s.%s", plural, v1beta2.GroupVersion.Group),
+		}, crdFromRuntime,
+	)
+
+	if k8serrors.IsNotFound(err) {
+		err = nil
```

Comment:
why not found, you assign nil to err?

Reply:
that's the problem, if crd not created, then the err is `nil`, it will return `crd, crdFromRuntime, nil`, but the `crdFromRuntime` is empty.
and in general, you should not mutate err, err should be treated as a readonly value.

Reply:
if the crd is not created, the error Is not nil. it is crd "x" not found.